src/teacup.coffee

```diff
@@ -43,6 +45,44 @@ merge_elements = (args...) ->
       result.push element unless element in result
   result
 
+# use queue for async rendering
+# anywhere htmlOut is modified we need to queue the update
+# opening tags go to htmlOut immediately
+# contents run immediately and delay further action until it's done
+# closing tags wait for contents to finish
+class Queue
+
+  constructor: (options) ->
+    @debug = false
+    @items = []
+    @names = []
+    @position = 0
+    @running = false
+
+  push: (name, fn) ->
+    @names.splice @position, 0, name
```

Comment:
names was simply for debugging, but I think maybe we could find a way to use it to get function names back in stack traces